### PR TITLE
refactor: peer connection handling

### LINKED ISSUE #4837: Mass reconnections when one mint is down
Using our hermes implementation to connect to mints to create p2pk ecash and I'm seeing a ton of logs within fedimint for a federation with one downed guardian. We do have to call subscribe for each operation, but does that try to retrigger reconnections? Or maybe this is just noisy logs and can be ignored? 

```
2024-04-05T04:21:09.120 app[d8d7457ae4e6e8] dfw [info] DEBUG jsonrpsee-client > Connecting to target: Target { sockaddrs: [185.193.127.171:443], host: "fedimintd.fedimint.downmint.com", host_header: "fedimintd.fedimint.downmint.com", _mode: Tls, path_and_query: "/" }

2024-04-05T04:21:09.209 app[d8d7457ae4e6e8] dfw [info] DEBUG fedimint_ln_client::receive > Still waiting preimage decryption for contract 43db1971184f64304d58d7b033c0fe1fd7bc5e9e1a8f723300e9860e2226db1d

2024-04-05T04:21:09.209 app[d8d7457ae4e6e8] dfw [info] DEBUG fedimint_ln_client::receive > Still waiting preimage decryption for contract 5665a4b80b1ced905fceef0634b4c65152c9cdebf80f8e468512e906f8eafc74

2024-04-05T04:21:09.210 app[d8d7457ae4e6e8] dfw [info] DEBUG fedimint_ln_client::receive > Still waiting preimage decryption for contract 749cc9add43cd0be70288f2c13fcecc7f07fbc9edbffe49cbee82e34a0abb929

2024-04-05T04:21:09.210 app[d8d7457ae4e6e8] dfw [info] DEBUG fedimint_ln_client::receive > Still waiting preimage decryption for contract 9da0882619d1d19aaaf9dd93419723cb80edc70207b486334968d724383fa888

2024-04-05T04:21:09.210 app[d8d7457ae4e6e8] dfw [info] DEBUG fedimint_ln_client::receive > Still waiting preimage decryption for contract d66d55e397ac1baead7bbbcc3ae71545172e0067edecd2bd407856e60c8ea43b

2024-04-05T04:21:09.233 app[d8d7457ae4e6e8] dfw [info] DEBUG jsonrpsee-client > Failed to connect to sockaddr: 185.193.127.171:443

2024-04-05T04:21:09.233 app[d8d7457ae4e6e8] dfw [info] DEBUG jsonrpsee-client > Connecting to target: Target { sockaddrs: [], host: "fedimintd.fedimint.downmint.com", host_header: "fedimintd.fedimint.downmint.com", _mode: Tls, path_and_query: "/" }

2024-04-05T04:21:09.234 app[d8d7457ae4e6e8] dfw [info] DEBUG jsonrpsee-client > Connecting to target: Target { sockaddrs: [], host: "fedimintd.fedimint.downmint.com", host_header: "fedimintd.fedimint.downmint.com", _mode: Tls, path_and_query: "/" }

2024-04-05T04:21:09.234 app[d8d7457ae4e6e8] dfw [info] DEBUG jsonrpsee-client > Connecting to target: Target { sockaddrs: [], host: "fedimintd.fedimint.downmint.com", host_header: "fedimintd.fedimint.downmint.com", _mode: Tls, path_and_query: "/" }

2024-04-05T04:21:09.234 app[d8d7457ae4e6e8] dfw [info] DEBUG jsonrpsee-client > Connecting to target: Target { sockaddrs: [], host: "fedimintd.fedimint.downmint.com", host_header: "fedimintd.fedimint.downmint.com", _mode: Tls, path_and_query: "/" }

2024-04-05T04:21:09.234 app[d8d7457ae4e6e8] dfw [info] DEBUG net::api > Unable to connect to peer peer_id=2 err=Error when opening the TCP socket: Connection refused (os error 111)

2024-04-05T04:21:09.234 app[d8d7457ae4e6e8] dfw [info] DEBUG jsonrpsee-client > Connecting to target: Target { sockaddrs: [185.193.127.171:443], host: "fedimintd.fedimint.downmint.com", host_header: "fedimintd.fedimint.downmint.com", _mode: Tls, path_and_query: "/" }

2024-04-05T04:21:09.363 app[d8d7457ae4e6e8] dfw [info] DEBUG jsonrpsee-client > Failed to connect to sockaddr: 185.193.127.171:443

2024-04-05T04:21:09.363 app[d8d7457ae4e6e8] dfw [info] DEBUG jsonrpsee-client > Connecting to target: Target { sockaddrs: [], host: "fedimintd.fedimint.downmint.com", host_header: "fedimintd.fedimint.downmint.com", _mode: Tls, path_and_query: "/" }

2024-04-05T04:21:09.363 app[d8d7457ae4e6e8] dfw [info] DEBUG jsonrpsee-client > Connecting to target: Target { sockaddrs: [], host: "fedimintd.fedimint.downmint.com", host_header: "fedimintd.fedimint.downmint.com", _mode: Tls, path_and_query: "/" }

2024-04-05T04:21:09.363 app[d8d7457ae4e6e8] dfw [info] DEBUG jsonrpsee-client > Connecting to target: Target { sockaddrs: [], host: "fedimintd.fedimint.downmint.com", host_header: "fedimintd.fedimint.downmint.com", _mode: Tls, path_and_query: "/" }

2024-04-05T04:21:09.363 app[d8d7457ae4e6e8] dfw [info] DEBUG jsonrpsee-client > Connecting to target: Target { sockaddrs: [], host: "fedimintd.fedimint.downmint.com", host_header: "fedimintd.fedimint.downmint.com", _mode: Tls, path_and_query: "/" }

2024-04-05T04:21:09.364 app[d8d7457ae4e6e8] dfw [info] DEBUG net::api > Unable to connect to peer peer_id=2 err=Error when opening the TCP socket: Connection refused (os error 111)

2024-04-05T04:21:09.364 app[d8d7457ae4e6e8] dfw [info] DEBUG jsonrpsee-client > Connecting to target: Target { sockaddrs: [185.193.127.171:443], host: "fedimintd.fedimint.downmint.com", host_header: "fedimintd.fedimint.downmint.com", _mode: Tls, path_and_query: "/" }
```

### PR DESCRIPTION
**Pull request by dpc**
See commit messages

Fix #4837

Summary:

To avoid very spamming with reconnection attempts, track connection attempts and add some delay.

Since `request` is cancellable, we could drop the future in the middle of connecting. Do it in a background task (Jit) instead.

### CODE REVIEW

**Comment by dpc** [comment]
I would rather land and follow up with any fixes due to #4940

File: fedimint-core/src/api.rs [hunk 1]
```diff
@@ -1062,11 +1063,119 @@ pub struct WsFederationApi<C = WsClient> {
...
+
+impl<C> FederationPeerClient<C>
+where
+    C: JsonRpcClient + 'static,
+{
+    pub fn new(peer_id: PeerId, url: SafeUrl) -> Self {
+        let shared: Arc<_> = tokio::sync::Mutex::new(FederationPeerClientShared::new()).into();
+
+        Self {
+            client: Self::new_jit_client(peer_id, url, shared.clone()),
```
[the remark this answers is not in the document]
Comment (elsirion) [reply fedimint-core/src/api.rs:1128]: Just connect asap. If it's the CLI we don't care about wasted resources imo as long as it doesn't impact latency. For anything more long running we'll want a connection anyway 99% of times.

File: fedimint-core/src/api.rs [hunk 2]
```diff
@@ -1670,163 +1663,6 @@ mod tests {
...
-        fed.request("", &[]).await.unwrap();
-        assert_eq!(
-            CONNECTION_COUNT.load(Ordering::SeqCst),
-            2,
-            "should connect again after disconnect"
-        );
-    }
-
-    #[test_log::test(tokio::test)]
-    async fn concurrent_requests() {
```
Comment (elsirion) [inline fedimint-core/src/api.rs:1658]: Testing if concurrency works seems kinda important.

File: fedimint-core/src/api.rs [hunk 3]
```diff
@@ -1516,13 +1516,26 @@ where
 {
     #[instrument(level = "trace", fields(peer = %self.peer_id, %method), skip_all)]
     pub async fn request(&self, method: &str, params: &[Value]) -> JsonRpcResult<Value> {
-        loop {
+        for attempts in 0.. {
             let rclient = self.client.read().await;
```
Comment (elsirion) [inline fedimint-core/src/api.rs:1520]: Could assert `attempts <= 1` here. Took me a bit to figure out that invariant.

**Review by elsirion — APPROVED** [review]
I strongly dislike reducing our unit testing coverage, but there's nothing broken about the code afaik. So ACK under the condition that we add back some testing for all that reconnect logic (especially concurrency-safety).

File: fedimint-core/src/api.rs [hunk 4]
```diff
@@ -1062,11 +1063,119 @@ pub struct WsFederationApi<C = WsClient> {
...
+    /// Wait (if needed) before reconnection attempt based on number of previous
+    /// attempts
+    async fn wait(&mut self) {
+        let desired_timeout = Duration::from_millis((self.connection_attempts * 100).min(5000));
+        let since_last_connect = now()
+            .duration_since(self.last_connection_attempt)
+            .unwrap_or_default();
+
+        let sleep_duration = desired_timeout.saturating_sub(since_last_connect);
+        if Duration::from_millis(0) < sleep_duration {
```
Comment (douglaz) [inline fedimint-core/src/api.rs:1091]:
```suggestion
        if sleep_duration < Duration::ZERO {
```

File: fedimint-core/src/api.rs [hunk 5]
```diff
@@ -1398,58 +1510,47 @@ pub struct PeerResponse<R> {
...
+        for attempts in 0.. {
+            let rclient = self.client.read().await;
+            match rclient.client.get_try().await {
+                Ok(client) if client.is_connected() => {
+                    return client.request::<_, _>(method, params).await;
+                }
+                Err(e) => {
+                    // Strategies using timeouts often depend on failing requests returning quickly,
+                    // so every request gets only one reconnection attempt.
+                    if 0 < attempts {
```
Comment (douglaz) [inline fedimint-core/src/api.rs:1528]: Else log something?

File: fedimint-core/src/api.rs [hunk 6]
```diff
@@ -1398,58 +1510,47 @@ pub struct PeerResponse<R> {
...
-                        // Low logging level because we will probably retry connecting later
-                        // we are going to retry, and a Federation peer being down is a fact
-                        // of life, and nothing to warn about right away
-                        debug!(
-                            target: LOG_NET_API,
-                            peer_id = %self.peer_id,
-                            %err, "Unable to connect to peer");
-                        return Err(err)?;
+                }
+                _ => {
```
Comment (douglaz) [inline fedimint-core/src/api.rs:1532]: is this `Ok(_client)`? Better be explicit about it.

File: fedimint-core/src/api.rs [hunk 7]
```diff
@@ -1398,58 +1510,47 @@ pub struct PeerResponse<R> {
...
+                    if 0 < attempts {
```
Comment (douglaz) [inline fedimint-core/src/api.rs:1533]: Else log something?

File: fedimint-core/src/api.rs [hunk 8]
```diff
@@ -1398,58 +1510,47 @@ pub struct PeerResponse<R> {
...
+                        )));
                     }
                 }
+            };
+
+            drop(rclient);
+            let mut wclient = self.client.write().await;
+            match wclient.client.get_try().await {
+                Ok(client) if client.is_connected() => {
+                    // someone else connected, just loop again
```
Comment (douglaz) [inline fedimint-core/src/api.rs:1545]: debug!/trace! ?

**Review by douglaz — APPROVED** [review]
Minor stuff, can be implemented later

**Comment by dpc** [comment]
All but extra tests in #4946

**Pull request merged**